libsentrykube/quickpatch.py

```diff
@@ -1,15 +1,142 @@
+    if "schema" not in patch_data:
+        raise ValueError(f"Schema not found in patch file {patch}.yaml")
+    if "patches" not in patch_data:
+        raise ValueError(f"Patches not found in patch file {patch}.yaml")
+
+    # Validate the schema
+    schema = patch_data.get("schema")
+    if (
+        "additionalProperties" not in schema
+        or schema["additionalProperties"] is not False
+    ):
+        raise ValueError(
+            f"Schema additionalProperties must be False in patch file {patch}.yaml"
+        )
+
+    # Find all variables enclosed in <> in the file content
+    file_content = file_path.read_text()
+    variables = set(re.findall(r"<\s*(\w+)\s*>", file_content))
+    # Remove 'resource' as it's a special case handled separately
+    variables.discard("resource")
+
+    # Get required fields from schema
+    required_fields = set(schema.get("required", []))
+
+    # Check if all variables are in required fields
+    missing_required = variables - required_fields
+    if missing_required:
+        raise ValueError(
+            f"Variables {missing_required} found in patch file but not listed in schema.required"
+        )
+
+    return patch_data
 
 
 def get_arguments(service: str, patch: str) -> Sequence[str]:
-    raise NotImplementedError
+    """
+    Returns the arguments required by the patch file
+    """
+    patch_file = find_patch_file(service, patch)
+    patch_data = load_and_validate_yaml(patch_file, patch)
+    return patch_data["schema"].get("required", [])
 
 
 def apply_patch(
-    service: str, resource: str, patch: str, arguments: Mapping[str, str]
+    service: str,
+    region: str,
+    resource: str,
+    patch: str,
+    arguments: MutableMapping[str, str],
+    cluster: str = "default",
 ) -> None:
     """
     Finds the patch file, the resource and applies the patch
-    TODO: Add parameters
+    Params:
+        service: The service to be patched
+        resource: The resource name to be patched
+        arguments: Arguments to be passed to the patch file and applied
     """
-    raise NotImplementedError
+    # Find files
+    patch_file = find_patch_file(service, patch)
+
+    # Check that the resource is allowed to be patched
+    patch_data = load_and_validate_yaml(patch_file, patch)
+    resource_mappings = patch_data.get("mappings", {})
+    if resource not in resource_mappings.keys():
+        raise ValueError(f"Resource {resource} is not allowed to be patched")
+
+    # Validate the arguments via jsonschema
+    schema = patch_data.get("schema")
+    try:
+        validate(instance=arguments, schema=schema)
+    except ValidationError as e:
+        raise ValidationError(f"Invalid arguments: {e.message}") from e
+
+    # Replace <resource_name> with the actual resource name
+    # Scan through the patch_data file and replace all matches of <resource_name>
+    # with the corresponding value in the arguments dictionary
+    variables = dict(arguments)
```

Comment:
Why turning it into a dict ?

Reply:
It's making a copy so that we don't mutate the passed in args